crates/metadata-store/src/rocksdb/server.rs

```diff
@@ -0,0 +1,344 @@
+                Self::log_error(self.get(key), &request).map_err(Into::into),
+            ),
+            RequestKind::GetVersion(GetVersion { key }) => ResponseKind::GetVersionResponse(
+                Self::log_error(self.get_version(key), &request).map_err(Into::into),
+            ),
+            RequestKind::Put(Put {
+                key,
+                value,
+                precondition,
+            }) => ResponseKind::PutResponse(
+                Self::log_error(self.put(key, value, precondition), &request).map_err(Into::into),
+            ),
+            RequestKind::Delete(Delete { key, precondition }) => ResponseKind::DeleteResponse(
+                Self::log_error(self.delete(key, precondition), &request).map_err(Into::into),
+            ),
+        };
+
+        let _ = task_center().spawn_child(
+            TaskKind::Disposable,
+            "send-metadata-store-response-to-client",
+            None,
+            {
+                let networking = self.networking.clone();
+                let request_id = *request_id;
+                async move {
+                    networking
+                        .send(
+                            caller,
+                            &MetadataStoreResponse {
+                                request_id,
+                                kind: response,
+                            },
+                        )
+                        .await?;
+                    Ok(())
+                }
+            },
+        );
+    }
+
+    fn get(&self, key: &ByteString) -> Result<Option<VersionedValue>> {
+        let cf_handle = self.kv_cf_handle();
+        let slice = self.db.get_pinned_cf(cf_handle, key)?;
+
+        if let Some(bytes) = slice {
+            Ok(Some(Self::decode(bytes)?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn get_version(&self, key: &ByteString) -> Result<Option<Version>> {
+        let cf_handle = self.kv_cf_handle();
+        let slice = self.db.get_pinned_cf(cf_handle, key)?;
+
+        if let Some(bytes) = slice {
+            let versioned_value = Self::decode::<VersionedValue>(bytes)?;
+            Ok(Some(versioned_value.version))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn put(
+        &self,
+        key: &ByteString,
+        value: &VersionedValue,
+        precondition: &Precondition,
+    ) -> Result<()> {
+        match precondition {
+            Precondition::None => Ok(self.write_versioned_kv_pair(key, value)?),
+            Precondition::DoesNotExist => {
+                let current_version = self.get_version(key)?;
+                if current_version.is_none() {
+                    Ok(self.write_versioned_kv_pair(key, value)?)
+                } else {
+                    Err(Error::kv_pair_exists())
+                }
+            }
+            Precondition::MatchesVersion(version) => {
+                let current_version = self.get_version(key)?;
```

Comment:
A future optimization could be to use merge operator instead.

Reply:
👍